sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/samples/count/CountSamples.groovy

```diff
@@ -30,65 +30,30 @@ class CountSamples implements CountSamplesInput{
-   * @since 1.0.0
-   */
   @Override
-  void countReceivedSamples(String projectCode) {
+  void countSamplePerStatus(String projectCode) {
     try {
       sampleStatuses = dataSource.fetchSampleStatusesForProject(projectCode)
+      int allSamples = sampleStatuses.size()
       // counts samples that have AT LEAST this status (or a later one)
       int receivedAmount = countSamplesFromStatus(Status.SAMPLE_RECEIVED)
-      output.countedReceivedSamples(projectCode, sampleStatuses.size(), receivedAmount)
+      output.countedReceivedSamples(projectCode,allSamples,receivedAmount)
+
+      int failedQc = sampleStatuses.findAll { it == Status.SAMPLE_QC_FAIL }.size()
+      output.countedFailedQcSamples(projectCode,allSamples,failedQc)
+
+      int libraryPrepFinished = countSamplesFromStatus(Status.LIBRARY_PREP_FINISHED)
+      output.countedLibraryPrepFinishedSamples(projectCode,allSamples,libraryPrepFinished)
+
+      int availableData = countSamplesFromStatus(Status.DATA_AVAILABLE)
+      output.countedAvailableSampleData(projectCode,allSamples,availableData)
     } catch (DataSourceException dataSourceException) {
       output.failedExecution(dataSourceException.getMessage())
     } catch (Exception ignored) {
       output.failedExecution("Could not count received samples.")
     }
   }
 
-  /**
-   * This method calls the output interface with the number of all samples in a project as
-   * well as the subset of samples which had a failed quality control
-   * In case of failure the output interface failure method is called.
-   * @since 1.0.0
-   */
-  @Override
-  void countQcFailedSamples(String projectCode) {
-    try {
-      sampleStatuses = dataSource.fetchSampleStatusesForProject(projectCode)
-      int receivedAmount = sampleStatuses.findAll { it == Status.SAMPLE_QC_FAIL }.size()
-      output.countedFailedQcSamples(projectCode, sampleStatuses.size(), receivedAmount)
-    }catch (Exception ignored) {
-      output.failedExecution("Could not count failed qc samples.")
-    }
-  }
-
-  @Override
-  void countAvailableDataSamples(String projectCode) {
-    try {
-      sampleStatuses = dataSource.fetchSampleStatusesForProject(projectCode)
-      int availableData = countSamplesFromStatus(Status.DATA_AVAILABLE)
-      output.countedAvailableSampleData(projectCode, sampleStatuses.size(), availableData)
-    } catch (Exception e) {
-      output.failedExecution(e.getMessage())
-    }
-  }
-
-  @Override
-  void countLibraryPrepFinishedSamples(String projectCode) {
-    try {
-      sampleStatuses = dataSource.fetchSampleStatusesForProject(projectCode)
-      int libraryPrepFinished = countSamplesFromStatus(Status.LIBRARY_PREP_FINISHED)
-      output.countedLibraryPrepFinishedSamples(projectCode, sampleStatuses.size(), libraryPrepFinished)
-    } catch (Exception e) {
-      output.failedExecution(e.getMessage())
-    }
-  }
-
   private int countSamplesFromStatus(Status status) {
```

Comment:
Since this method contains business logic that is agnostic to the current state, you can make this method static and add the list as an input argument. This would make sure the logic returns the same value, regardless of when it is called.
```suggestion
  private static int countSamplesFromStatus(List<Status> sampleStatuses, Status status) {
```

Reply:
I don't like adding more parameters to the method but how about making the list static along with this method and keeping the number of parameters as is?

Reply:
That would make the content of the list static. We should not do that because we would modify all instances of the use case. My point was rather to make the logic static with the class values, this would not be achieved by making the field static and would lead to different behaviour. I think that two parameters are still very much acceptable.